columnflow/production/cms/dy.py

```diff
@@ -0,0 +1,501 @@
+    for column_name, syst in weights_list:
+        variable_map_syst = {**variable_map, "syst": syst}
+
+        # evaluating dy weights given a certain era, ptll array and sytematic shift
+        inputs = [variable_map_syst[inp.name] for inp in self.dy_corrector.inputs]
+        dy_weight = self.dy_corrector.evaluate(*inputs)
+
+        # save the weights in a new column
+        events = set_ak_column(events, column_name, dy_weight, value_type=np.float32)
+
+    return events
+
+
+@dy_weights.init
+def dy_weights_init(self: Producer) -> None:
+    # the number of weights in partial run 3 is always 10
+    if self.config_inst.campaign.x.year not in (2022, 2023):
+        raise NotImplementedError(
+            f"campaign year {self.config_inst.campaign.x.year} is not yet supported by {self.cls_name}",
+        )
+    self.n_unc = 10
+
+    # register dynamically produced weight columns
+    for i in range(self.n_unc):
+        self.produces.add(f"dy_weight{i + 1}_{{up,down}}")
+
+
+@dy_weights.requires
+def dy_weights_requires(self: Producer, reqs: dict) -> None:
+    """
+    Adds the requirements needed the underlying task to derive the Drell-Yan weights into *reqs*.
+    """
+    if "external_files" in reqs:
+        return
+
+    from columnflow.tasks.external import BundleExternalFiles
+
+    reqs["external_files"] = BundleExternalFiles.req(self.task)
+
+
+@dy_weights.setup
+def dy_weights_setup(
+    self: Producer,
+    reqs: dict,
+    inputs: dict,
+    reader_targets: InsertableDict,
+) -> None:
+    """
+    Loads the Drell-Yan weight calculator from the external files bundle and saves them in the
+    py:attr:`dy_corrector` attribute for simpler access in the actual callable. The number of uncertainties
+    is calculated, per era, by another correcter in the external file and is saved in the
+    py:attr:`dy_unc_corrector` attribute.
+    """
+    bundle = reqs["external_files"]
+
+    # import all correctors from the external file
+    import correctionlib
+
+    correctionlib.highlevel.Correction.__call__ = (
+        correctionlib.highlevel.Correction.evaluate
+    )
+    correction_set = correctionlib.CorrectionSet.from_string(
+        self.get_dy_weight_file(bundle.files).load(formatter="gzip").decode("utf-8"),
+    )
+
+    # check number of fetched correctors
+    if len(correction_set.keys()) != 2:
+        raise Exception("Expected exactly two types of Drell-Yan correction")
+
+    # create the weight and uncertainty correctors
+    self.dy_config: DrellYanConfig = self.get_dy_weight_config()
+    self.dy_corrector = correction_set[self.dy_config.correction]
+    self.dy_unc_corrector = correction_set[self.dy_config.unc_correction]
+
+    dy_n_unc = int(self.dy_unc_corrector.evaluate(self.dy_config.order))
+
+    if dy_n_unc != self.n_unc:
+        raise ValueError(
+            f"Expected {self.n_unc} uncertainties, got {dy_n_unc}",
+        )
+
+
+#
+# Helper functions for kinematics.
+#
+def GetXYfromPTPHI(pt, phi):
```

Comment:
I just wanted to start reviewing this, but I stumbled over the four-vector logic here and below. Even though they look good actually, I think we should stick to the existing methods available through the nano behavior. Or was there an issue with the behavior we should solve?

Reply:
I'll change these today. This was my first contact point with implemeting columnar analysis. Now I have some more experience from BTV workflows in coffea and can update these to nano behaviour.

Reply:
I could not push to this branch (on the columnflow project), so I made a [MR targeting this branch](https://github.com/columnflow/columnflow/compare/tmp_dy...philippgadow:columnflow:tmp_dy?expand=1) on my project.

Reply:
I now pushed directly to this branch.